File: packages/core/src/pipeline-variables.ts
/**
 * @fileoverview Pipeline variable references, as specified in `spec/pipeline-variables.md`.
 *
 * A step input string is a reference only when the whole string is one of `$prev`, `$first`,
 * `$steps[N]`, `$steps.<alias>` or `$input`, optionally followed by `.<path>`. Every other string
 * is a literal, and a leading `$$` sends a literal `$`. Paths reach only own keys of plain JSON
 * objects and in-bounds array indices, never prototypes, getters or `__`-prefixed keys. A
 * reference that cannot be resolved is absent: omitted from objects, `null` in arrays, and never
 * equal to anything in a `when` condition.
 *
 * Separated from `pipeline.ts` to keep both modules within the repository file-size convention.
 */

import type { PipelineCondition, PipelineContext } from './pipeline.js';

/** Step inputs and the request `input` may nest at most this many objects and arrays. */
export const MAX_NESTING_DEPTH = 64;

/** Strings longer than this are literals and are never resolved. */
export const MAX_REFERENCE_LENGTH = 1024;

type Segment = { key: string; index?: number };
type Source =
	| { kind: 'prev' | 'first' | 'input' }
	| { kind: 'index'; index: number }
	| { kind: 'alias'; alias: string };
type Reference = { source: Source; path: Segment[] };

/** A key (no `.`, `[`, `]` or whitespace), optionally followed by one `[N]` index. */
const SEGMENT_PATTERN = /^([^.[\]\s]+)(?:\[(\d+)\])?$/;
const STEP_INDEX_PATTERN = /^\$steps\[(\d+)\]/;
const ALIAS_PATTERN = /^[^.[\]\s]+$/;
const DIGITS_PATTERN = /^\d+$/;
const NAMED_SOURCES = ['$prev', '$first', '$input'] as const;

/** True for objects whose prototype is `Object.prototype` or `null` (what JSON produces). */
export function isPlainObject(value: unknown): value is Record<string, unknown> {
	if (typeof value !== 'object' || value === null || Array.isArray(value)) return false;
	const prototype: unknown = Object.getPrototypeOf(value);
	return prototype === Object.prototype || prototype === null;
}

/**
 * Copy step data so that steps never share objects.
 *
 * Uses `structuredClone`. Data that cannot be structured-cloned (for example, objects holding
 * functions) falls back to its JSON view; data with no JSON view either is returned as is.
 */
export function copyPipelineData(value: unknown): unknown {
	if (typeof value !== 'object' || value === null) return value;
	try {
		return structuredClone(value);
	} catch {
		try {
			const json = JSON.stringify(value);
			return json === undefined ? undefined : JSON.parse(json);
		} catch {
			return value;
		}
	}
}

function parsePath(path: string): Segment[] | undefined {
	const segments: Segment[] = [];
	for (const part of path.split('.')) {
		const match = SEGMENT_PATTERN.exec(part);
		const key = match?.[1];
		if (match === null || key === undefined) return undefined;
		const index = match[2];
		segments.push(index === undefined ? { key } : { key, index: Number(index) });
	}
	return segments;
}

/** Parse a reference string, or return undefined when the string is a literal. */
function parseReference(ref: string): Reference | undefined {
	if (ref.length > MAX_REFERENCE_LENGTH || !ref.startsWith('$') || ref.startsWith('$$')) {
		return undefined;
	}
	let source: Source;
	let rest: string;
	const indexMatch = STEP_INDEX_PATTERN.exec(ref);
	const named = NAMED_SOURCES.find((name) => ref === name || ref.startsWith(`${name}.`));
	if (indexMatch !== null && indexMatch[1] !== undefined) {
		source = { kind: 'index', index: Number(indexMatch[1]) };
		rest = ref.slice(indexMatch[0].length);
	} else if (ref.startsWith('$steps.')) {
		const body = ref.slice('$steps.'.length);
		const dot = body.indexOf('.');
		const alias = dot < 0 ? body : body.slice(0, dot);
		if (!ALIAS_PATTERN.test(alias)) return undefined;
		source = { kind: 'alias', alias };
		rest = dot < 0 ? '' : body.slice(dot);
	} else if (named !== undefined) {
		source = { kind: named === '$prev' ? 'prev' : named === '$first' ? 'first' : 'input' };
		rest = ref.slice(named.length);
	} else {
		return undefined;
	}
	if (rest === '') return { source, path: [] };
	if (!rest.startsWith('.')) return undefined;
	const path = parsePath(rest.slice(1));
	return path === undefined ? undefined : { source, path };
}

/** The value of an own data property, without invoking getters. */
function ownData(container: object, key: string): unknown {
	if (!Object.hasOwn(container, key)) return undefined;
	const descriptor = Object.getOwnPropertyDescriptor(container, key);
	return descriptor !== undefined && 'value' in descriptor ? descriptor.value : undefined;
}

function element(array: readonly unknown[], index: number): unknown {
	return Number.isSafeInteger(index) && index < array.length
		? ownData(array, String(index))
		: undefined;
}

function child(container: unknown, key: string): unknown {
	if (key.startsWith('__')) return undefined;
	if (Array.isArray(container)) {
		return DIGITS_PATTERN.test(key) ? element(container, Number(key)) : undefined;
	}
	return isPlainObject(container) ? ownData(container, key) : undefined;
}

function traverse(root: unknown, path: readonly Segment[]): unknown {
	let current = root;
	for (const { key, index } of path) {
		current = child(current, key);
		if (index !== undefined) current = Array.isArray(current) ? element(current, index) : undefined;
		if (current === undefined) return undefined;
	}
	return current;
}

function sourceData(source: Source, context: PipelineContext): unknown {
	switch (source.kind) {
		case 'prev':
			return context.previousResult?.data;
		case 'first':
			return context.steps[0]?.data;
		case 'input':
			return context.pipelineInput;
		case 'index':
			return Number.isSafeInteger(source.index) && source.index < context.steps.length
				? context.steps[source.index]?.data
				: undefined;
		case 'alias':
			if (source.alias.startsWith('__')) return undefined;
			return context.steps.find((step) => step.alias === source.alias)?.data;
	}
}

function lookup(ref: Reference, context: PipelineContext): unknown {
	return traverse(sourceData(ref.source, context), ref.path);
}

/**
 * Resolve one step input string against the pipeline context.
 *
 * - A reference (`$prev`, `$first`, `$steps[N]`, `$steps.<alias>`, `$input`, each optionally
 *   followed by `.<path>`) resolves to that data, or to `undefined` (absent) when it cannot be
 *   resolved: unknown alias, index out of bounds, missing key, skipped or failed step, a
 *   `__`-prefixed segment, or no request `input`.
 * - A string starting with `$$` is a literal with one leading `$` removed.
 * - Any other string (`$9.99`, `$HOME`, `$prevx`, or anything longer than 1024 characters) is
 *   returned unchanged.
 *
 * `$input` is the pipeline request's own `input` field, never the host's execution context.
 * A path follows only own keys of plain JSON objects and in-bounds array indices; a purely
 * numeric segment (`items.2`) indexes an array. Prototype properties such as `constructor`
 * never match.
 *
 * @param ref - A step input string (e.g. '$prev.user.name', '$steps.order.items[0]', '$$literal')
 * @param context - Pipeline execution context
 * @returns The resolved value, the literal string, or undefined when a reference is unresolved
 *
 * @example
 * ```typescript
 * resolveVariable('$prev.user.name', context); // data of the previous step
 * resolveVariable('$steps.order.items[0]', context); // first item of the step aliased 'order'
 * resolveVariable('$9.99', context); // '$9.99' (not a reference)
 * resolveVariable('$$prev', context); // '$prev' (escaped literal)
 * ```
 */
export function resolveVariable(ref: string, context: PipelineContext): unknown {
	if (ref.startsWith('$$')) return ref.slice(1);
	const parsed = parseReference(ref);
	return parsed === undefined ? ref : lookup(parsed, context);
}

/** Marks a reference that could not be resolved, as opposed to a literal `undefined`. */
const ABSENT: unique symbol = Symbol('absent');

function resolveValue(value: unknown, context: PipelineContext, depth: number): unknown {
	if (typeof value === 'string') {
		if (value.startsWith('$$')) return value.slice(1);
		const parsed = parseReference(value);
		if (parsed === undefined) return value;
		const resolved = lookup(parsed, context);
		return resolved === undefined ? ABSENT : copyPipelineData(resolved);
	}
	const isArray = Array.isArray(value);
	if (!isArray && !isPlainObject(value)) return value;
	if (depth >= MAX_NESTING_DEPTH) {
		throw new Error(`Pipeline step input is nested deeper than ${MAX_NESTING_DEPTH} levels`);
	}
	if (isArray) {
		return value.map((item) => {
			const resolved = resolveValue(item, context, depth + 1);
			return resolved === ABSENT ? null : resolved;
		});
	}
	const entries: Array<[string, unknown]> = [];
	for (const [key, item] of Object.entries(value)) {
		const resolved = resolveValue(item, context, depth + 1);
		if (resolved !== ABSENT) entries.push([key, resolved]);
	}
	return Object.fromEntries(entries);
}

/**
 * Resolve every reference in a step input (see {@link resolveVariable}).
 *
 * Walks arrays and plain objects; other values pass through unchanged. An unresolved reference
 * is omitted as an object property and becomes `null` as an array element. Resolved values are
 * copies, so a handler that mutates its input cannot change another step's data.
 *
 * @param input - Step input potentially containing references
 * @param context - Pipeline execution context
 * @returns The input with all references resolved
 * @throws Error when arrays and objects nest deeper than 64 levels; `executePipeline` rejects
 *   such inputs with `VALIDATION_ERROR` before any step runs
 */
export function resolveVariables(input: unknown, context: PipelineContext): unknown {
	const resolved = resolveValue(input, context, 0);
	return resolved === ABSENT ? undefined : resolved;
}

/**
 * Get a nested value from plain JSON data using dot notation.
 *
 * Only own keys of plain objects and in-bounds array indices are followed; `__`-prefixed keys,
 * prototype properties and getters are never read.
 *
 * @param obj - The data to traverse
 * @param path - Dot-separated path (e.g., 'user.profile.name', 'items[1]', 'items.1')
 * @returns The value at the path, or undefined if not found
 *
 * @example
 * ```typescript
 * getNestedValue({ user: { name: 'Alice' } }, 'user.name'); // => 'Alice'
 * getNestedValue({ items: [1, 2, 3] }, 'items[1]'); // => 2
 * getNestedValue({}, 'constructor'); // => undefined
 * ```
 */
export function getNestedValue(obj: unknown, path: string): unknown {
	const segments = parsePath(path);
	return segments === undefined ? undefined : traverse(obj, segments);
}

/** A condition operand: a reference, or absent when it is not one or cannot be resolved. */
function operand(ref: unknown, context: PipelineContext): unknown {
	if (typeof ref !== 'string') return undefined;
	const parsed = parseReference(ref);
	return parsed === undefined ? undefined : lookup(parsed, context);
}

function compare(
	pair: readonly [unknown, unknown],
	context: PipelineContext,
	test: (value: unknown, expected: unknown) => boolean
): boolean {
	const value = operand(pair[0], context);
	const expected = pair[1];
	return value !== undefined && expected !== undefined && test(value, expected);
}

/** Structural equality of JSON values: objects and arrays compare by content, never identity. */
function jsonEqual(a: unknown, b: unknown): boolean {
	if (a === b) return true;
	if (Array.isArray(a) || Array.isArray(b)) {
		return (
			Array.isArray(a) &&
			Array.isArray(b) &&
			a.length === b.length &&
			a.every((item, i) => jsonEqual(item, b[i]))
		);
	}
	if (typeof a !== 'object' || typeof b !== 'object' || a === null || b === null) return false;
	const aKeys = Object.keys(a);
	const bKeys = Object.keys(b);
	return (
		aKeys.length === bKeys.length &&
		aKeys.every(
			(key) => Object.hasOwn(b, key) && jsonEqual(Reflect.get(a, key), Reflect.get(b, key))
		)
	);
}

function numeric(value: unknown, expected: unknown, test: (a: number, b: number) => boolean) {
	return typeof value === 'number' && typeof expected === 'number' && test(value, expected);
}

/**
 * Evaluate a pipeline condition against the current context.
 *
 * The first operand of every comparison is a reference. A string that is not a reference, or a
 * reference that cannot be resolved, is absent: `$exists` is false (as it is for `null`), and
 * `$eq`, `$ne`, `$gt`, `$gte`, `$lt` and `$lte` are false when either operand is absent. `$eq` and
 * `$ne` compare JSON values structurally.
 *
 * @param condition - The condition to evaluate
 * @param context - Pipeline execution context
 * @returns true if the condition is met, false otherwise
 */
export function evaluateCondition(condition: PipelineCondition, context: PipelineContext): boolean {
	if ('$exists' in condition) {
		const value = operand(condition.$exists, context);
		return value !== undefined && value !== null;
	}
	if ('$eq' in condition) return compare(condition.$eq, context, jsonEqual);
	if ('$ne' in condition) return compare(condition.$ne, context, (a, b) => !jsonEqual(a, b));
	if ('$gt' in condition) {
		return compare(condition.$gt, context, (a, b) => numeric(a, b, (x, y) => x > y));
	}
	if ('$gte' in condition) {
		return compare(condition.$gte, context, (a, b) => numeric(a, b, (x, y) => x >= y));
	}
	if ('$lt' in condition) {
		return compare(condition.$lt, context, (a, b) => numeric(a, b, (x, y) => x < y));
	}
	if ('$lte' in condition) {
		return compare(condition.$lte, context, (a, b) => numeric(a, b, (x, y) => x <= y));
	}
	if ('$and' in condition) return condition.$and.every((c) => evaluateCondition(c, context));
	if ('$or' in condition) return condition.$or.some((c) => evaluateCondition(c, context));
	if ('$not' in condition) return !evaluateCondition(condition.$not, context);
	return false;
}

/**
 * Alias for resolveVariable (backwards compatibility).
 *
 * @deprecated Use resolveVariable instead
 */
export const resolveReference = resolveVariable;

File: python/src/afd/core/pipeline_variables.py
"""Pipeline variable references, as specified in ``spec/pipeline-variables.md``.

A step input value is a reference only when the whole string is one of:

- ``$prev`` / ``$prev.<path>``: data of the previous successful step;
- ``$first`` / ``$first.<path>``: data of the first step;
- ``$steps[N]`` / ``$steps[N].<path>``: data of step ``N`` (0-based);
- ``$steps.<alias>`` / ``$steps.<alias>.<path>``: data of the step named ``<alias>``;
- ``$input`` / ``$input.<path>``: the pipeline request's ``input``.

A key is any run of characters other than ``.``, ``[``, ``]`` and whitespace,
optionally followed by one ``[N]`` index. Any other string that starts with
``$`` (``$9.99``, ``$HOME``, ``$prevx``, ``$steps[0][1]``) is a literal, a
string starting with ``$$`` is a literal with one ``$`` removed (at any
length), and would-be references longer than :data:`MAX_REFERENCE_LENGTH`
characters are literals.

Resolution only follows JSON data: step data is first converted to its JSON
representation (pydantic models and dataclasses included), a key segment only
matches an own key of a JSON object, a segment starting with ``__`` never
matches, and an index must be in bounds. Attributes are never read.

A reference that cannot be resolved is *absent* (:data:`ABSENT`): the property
is omitted from an object, and becomes ``None`` in a list.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Any, Callable, Optional, Tuple

from afd.core.wire import to_wire

MAX_INPUT_DEPTH = 64
"""Maximum nesting depth of a step input or the request input (the outermost
object or array is level 1)."""

MAX_REFERENCE_LENGTH = 1024
"""Strings longer than this are never resolved as references."""


class _Absent:
    """Sentinel for an unresolved reference."""

    _instance: Optional["_Absent"] = None

    def __new__(cls) -> "_Absent":
        if cls._instance is None:
            cls._instance = super().__new__(cls)
        return cls._instance

    def __repr__(self) -> str:
        return "ABSENT"

    def __bool__(self) -> bool:
        return False


ABSENT: Any = _Absent()
"""The value of a reference that cannot be resolved."""

# A key is any non-empty run of characters other than '.', '[', ']' and whitespace.
_KEY = r"[^.\[\]\s]+"
_SEGMENT_RE = re.compile(rf"({_KEY})(?:\[([0-9]+)\])?")
_REFERENCE_RE = re.compile(
    rf"\$(?:(?P<root>prev|first|input)|steps\[(?P<index>[0-9]+)\]|steps\.(?P<alias>{_KEY}))"
    r"(?:\.(?P<path>.+))?",
    re.DOTALL,
)

PathSegment = Tuple[str, Optional[int]]
"""A path segment: a key, optionally followed by an array index (``items[2]``)."""


@dataclass(frozen=True)
class VariableReference:
    """A parsed variable reference.

    Attributes:
        root: ``"prev"``, ``"first"``, ``"input"`` or ``"steps"``.
        index: Step index for ``$steps[N]``.
        alias: Step alias for ``$steps.<alias>``.
        path: Path segments after the root.
    """

    root: str
    index: Optional[int] = None
    alias: Optional[str] = None
    path: Tuple[PathSegment, ...] = ()


def parse_path(path: str) -> Optional[Tuple[PathSegment, ...]]:
    """Parse a dotted path into segments, or return None if it is malformed."""
    segments = []
    for part in path.split("."):
        match = _SEGMENT_RE.fullmatch(part)
        if match is None:
            return None
        index = match.group(2)
        segments.append((match.group(1), int(index) if index is not None else None))
    return tuple(segments)


def parse_reference(value: str) -> Optional[VariableReference]:
    """Parse a reference string.

    Returns:
        The reference, or None when the string is a literal (it does not match
        a reference form, starts with ``$$`` or is too long).
    """
    if len(value) > MAX_REFERENCE_LENGTH:
        return None
    match = _REFERENCE_RE.fullmatch(value)
    if match is None:
        return None

    path: Tuple[PathSegment, ...] = ()
    if match.group("path") is not None:
        parsed = parse_path(match.group("path"))
        if parsed is None:
            return None
        path = parsed

    if match.group("root") is not None:
        return VariableReference(root=match.group("root"), path=path)
    if match.group("index") is not None:
        return VariableReference(root="steps", index=int(match.group("index")), path=path)
    return VariableReference(root="steps", alias=match.group("alias"), path=path)


def json_view(value: Any) -> Any:
    """Return the JSON representation that resolution traverses.

    Returns :data:`ABSENT` for data that cannot be represented (for example,
    nesting too deep to serialize).
    """
    try:
        return to_wire(value)
    except ValueError:
        return ABSENT


def _index(current: Any, index: int) -> Any:
    if isinstance(current, list) and index < len(current):
        return current[index]
    return ABSENT


def follow_path(view: Any, path: Tuple[PathSegment, ...]) -> Any:
    """Follow path segments through JSON data (dicts, lists and scalars only).

    Returns:
        The value at the path, or :data:`ABSENT`.
    """
    current = view
    for key, index in path:
        if key.startswith("__"):
            return ABSENT
        if isinstance(current, dict):
            if key not in current:
                return ABSENT
            current = current[key]
        elif isinstance(current, list) and key.isascii() and key.isdigit():
            current = _index(current, int(key))
        else:
            return ABSENT
        if current is ABSENT:
            return ABSENT
        if index is not None:
            current = _index(current, index)
            if current is ABSENT:
                return ABSENT
    return current


def get_nested_value(obj: Any, path: str) -> Any:
    """Get a nested value from data using a dotted path.

    The data is viewed as JSON first (pydantic models and dataclasses by their
    serialized form), so attributes are never read and a segment starting with
    ``__`` never resolves.

    Returns:
        The value, or None when the path cannot be resolved.

    Example:
        >>> get_nested_value({"user": {"tags": ["a", "b"]}}, "user.tags[1]")
        'b'
        >>> get_nested_value({"user": {}}, "user.__class__") is None
        True
    """
    if obj is None:
        return None
    segments = parse_path(path)
    if segments is None:
        return None
    value = follow_path(json_view(obj), segments)
    return None if value is ABSENT else value


def unescape_literal(value: str) -> str:
    """Remove the escape from a ``$$``-prefixed literal."""
    return value[1:] if value.startswith("$$") else value


RootLookup = Callable[[VariableReference], Any]
"""Returns the JSON view of a reference's root data, or :data:`ABSENT`."""


def resolve_string(value: str, lookup: RootLookup) -> Any:
    """Resolve one string: a reference, an escaped literal, or a literal.

    Returns:
        The resolved value, the literal string, or :data:`ABSENT`.
    """
    if not value.startswith("$"):
        return value
    if value.startswith("$$"):
        return value[1:]
    reference = parse_reference(value)
    if reference is None:
        return value
    root = lookup(reference)
    if root is ABSENT:
        return ABSENT
    return follow_path(root, reference.path)


def resolve_input(value: Any, lookup: RootLookup) -> Any:
    """Resolve references anywhere in a step input.

    Unresolved references are omitted from objects and become ``None`` in lists.
    A top-level unresolved reference returns :data:`ABSENT`.
    """
    if isinstance(value, str):
        return resolve_string(value, lookup)
    if isinstance(value, dict):
        resolved = {}
        for key, item in value.items():
            item_value = resolve_input(item, lookup)
            if item_value is not ABSENT:
                resolved[key] = item_value
        return resolved
    if isinstance(value, (list, tuple)):
        items = []
        for item in value:
            item_value = resolve_input(item, lookup)
            items.append(None if item_value is ABSENT else item_value)
        return items
    return value


def exceeds_depth(value: Any, limit: int = MAX_INPUT_DEPTH) -> bool:
    """Return True if containers nest deeper than ``limit`` levels.

    The outermost container is level 1. Recursion stops one level past the
    limit, so hostile input cannot exhaust the stack.
    """

    def walk(item: Any, depth: int) -> bool:
        if isinstance(item, dict):
            children: Any = item.values()
        elif isinstance(item, (list, tuple)):
            children = item
        else:
            return False
        if depth > limit:
            return True
        return any(walk(child, depth + 1) for child in children)

    return walk(value, 1)


def is_number(value: Any) -> bool:
    """Return True for JSON numbers (booleans excluded)."""
    return isinstance(value, (int, float)) and not isinstance(value, bool)


def json_equals(left: Any, right: Any) -> bool:
    """Compare JSON values the way JavaScript's ``===`` compares scalars.

    Booleans never equal numbers, and containers compare by value.
    """
    if isinstance(left, bool) or isinstance(right, bool):
        return isinstance(left, bool) and isinstance(right, bool) and left == right
    if is_number(left) or is_number(right):
        return is_number(left) and is_number(right) and left == right
    if isinstance(left, dict) and isinstance(right, dict):
        return left.keys() == right.keys() and all(
            json_equals(left[key], right[key]) for key in left
        )
    if isinstance(left, (list, tuple)) and isinstance(right, (list, tuple)):
        return len(left) == len(right) and all(
            json_equals(a, b) for a, b in zip(left, right, strict=True)
        )
    return type(left) is type(right) and left == right
